File: examples/cpp/Depth/focused_depth.cpp
/**
 * Focused Depth demo.
 *
 * Runs an object-detection network on the color camera and feeds the resulting
 * ImgDetections into the Depth node's inputDetections. The Depth node computes a
 * focused depth map (full frame, only the detected regions filled) and a matching
 * confidence map.
 *
 * Works on RVC4 devices. The ImgDetections are automatically remapped to the left
 * stereo frame using the ImgDetections transformation when available.
 */

#include <algorithm>
#include <argparse/argparse.hpp>
#include <iostream>
#include <opencv2/opencv.hpp>
#include <optional>
#include <string>

#include "depthai/common/CameraBoardSocket.hpp"
#include "depthai/common/CameraSensorType.hpp"
#include "depthai/depthai.hpp"
#include "depthai/modelzoo/Zoo.hpp"
#include "depthai/pipeline/node/Camera.hpp"
#include "depthai/pipeline/node/Depth.hpp"
#include "depthai/pipeline/node/DetectionNetwork.hpp"

namespace {

cv::Mat colorizeDepth(const cv::Mat& frameDepth) {
    if(frameDepth.empty() || frameDepth.channels() != 1) {
        return cv::Mat::zeros(frameDepth.size(), CV_8UC3);
    }

    cv::Mat depth32f;
    frameDepth.convertTo(depth32f, CV_32F);

    const cv::Mat nonZeroMask = depth32f != 0.0f;
    const int nz = cv::countNonZero(nonZeroMask);
    if(nz == 0) {
        return cv::Mat::zeros(frameDepth.size(), CV_8UC3);
    }

    std::vector<float> values;
    values.reserve(static_cast<size_t>(nz));
    for(int r = 0; r < depth32f.rows; ++r) {
        const float* d = depth32f.ptr<float>(r);
        const uchar* m = nonZeroMask.ptr<uchar>(r);
        for(int c = 0; c < depth32f.cols; ++c) {
            if(m[c]) {
                values.push_back(d[c]);
            }
        }
    }

    std::sort(values.begin(), values.end());
    auto pct = [&](double p) {
        const size_t idx = static_cast<size_t>(std::round((p / 100.0) * (values.size() - 1)));
        return values[idx];
    };

    const float minDepth = pct(3.0);
    const float maxDepth = pct(95.0);

    cv::Mat logDepth;
    depth32f.copyTo(logDepth);
    logDepth.setTo(minDepth, ~nonZeroMask);
    cv::log(logDepth, logDepth);

    const float logMinDepth = std::log(minDepth);
    const float logMaxDepth = std::log(maxDepth);

    cv::min(logDepth, logMaxDepth, logDepth);
    cv::max(logDepth, logMinDepth, logDepth);
    logDepth = (logDepth - logMinDepth) * (255.0f / (logMaxDepth - logMinDepth));

    cv::Mat depth8U;
    logDepth.convertTo(depth8U, CV_8U);

    cv::Mat depthFrameColor;
    cv::applyColorMap(depth8U, depthFrameColor, cv::COLORMAP_JET);
    depthFrameColor.setTo(cv::Scalar::all(0), ~nonZeroMask);
    return depthFrameColor;
}

cv::Mat colorizeConfidence(const cv::Mat& frame) {
    if(frame.empty() || frame.channels() != 1) {
        return cv::Mat::zeros(frame.size(), CV_8UC3);
    }

    cv::Mat normalized;
    if(frame.depth() == CV_16U) {
        double maxValue = 0.0;
        cv::minMaxLoc(frame, nullptr, &maxValue);
        if(maxValue <= 0.0) {
            return cv::Mat::zeros(frame.size(), CV_8UC3);
        }
        cv::Mat scaled;
        frame.convertTo(scaled, CV_32F, 255.0 / maxValue);
        scaled.convertTo(normalized, CV_8U);
    } else {
        frame.convertTo(normalized, CV_8U);
    }

    cv::Mat colorized;
    cv::applyColorMap(normalized, colorized, cv::COLORMAP_JET);
    return colorized;
}

}  // namespace

int main(int argc, char** argv) {
    argparse::ArgumentParser program("focused_depth", "1.0.0");
    program.add_description(
        "Focused Depth demo.\n\n"
        "Runs an object-detection network on the color camera and feeds the resulting "
        "ImgDetections into the Depth node's inputDetections. The Depth node computes a "
        "focused depth map (full frame, only the detected regions filled) and a matching "
        "confidence map.");
    program.add_argument("--model").default_value(std::string("yolov6-nano")).help("Detection model description (default: yolov6-nano)");
    program.add_argument("--fps").scan<'g', float>().default_value(30.0f).help("Requested camera FPS");

    try {
        program.parse_args(argc, argv);
    } catch(const std::runtime_error& err) {
        std::cerr << err.what() << '\n';
        std::cerr << program;
        return EXIT_FAILURE;
    }

    const std::string model = program.get<std::string>("--model");
    const float fps = program.get<float>("--fps");

    dai::Pipeline pipeline;
    const auto device = pipeline.getDefaultDevice();
    if(device == nullptr) {
        throw std::runtime_error("No default device available.");
    }

    std::optional<dai::CameraBoardSocket> colorSocket;
    for(const auto& features : device->getConnectedCameraFeatures()) {
        if(std::find(features.supportedTypes.begin(), features.supportedTypes.end(), dai::CameraSensorType::COLOR) != features.supportedTypes.end()) {
            colorSocket = features.socket;
            break;
        }
    }
    if(!colorSocket.has_value()) {
        throw std::runtime_error("No color camera found on this device for object detection.");
    }
    std::cout << "Running detection on color camera socket: " << static_cast<int>(*colorSocket) << std::endl;

    auto camRgb = pipeline.create<dai::node::Camera>();
    camRgb->build(*colorSocket, std::nullopt, fps);

    auto detectionNetwork = pipeline.create<dai::node::DetectionNetwork>();
    detectionNetwork->build(camRgb, dai::NNModelDescription{model});
    detectionNetwork->setConfidenceThreshold(0.25f);

    auto depthNode = pipeline.create<dai::node::Depth>();
    depthNode->setFocusModels({dai::DeviceModelZoo::NEURAL_DEPTH_192X120});
    depthNode->setFocusSelectionMode(dai::node::FocusController::SelectionMode::LARGEST);
    depthNode->build(fps);
    detectionNetwork->out.link(depthNode->inputDetections);

    auto rgbQueue = detectionNetwork->passthrough.createOutputQueue(4, false);
    auto detQueue = detectionNetwork->out.createOutputQueue(4, false);
    auto focusedDepthQueue = depthNode->focusedDepth().createOutputQueue(4, false);
    auto focusedConfQueue = depthNode->focusedConfidence().createOutputQueue(4, false);

    pipeline.build();

    std::cout << "Focused depth: largest object, neural model 192x120, requested FPS=" << fps << '\n';

    pipeline.start();

    while(pipeline.isRunning()) {
        auto frameRgb = rgbQueue->get<dai::ImgFrame>();
        auto detections = detQueue->get<dai::ImgDetections>();
        auto frameFocusedDepth = focusedDepthQueue->get<dai::ImgFrame>();
        auto frameFocusedConf = focusedConfQueue->get<dai::ImgFrame>();

        if(frameRgb != nullptr && detections != nullptr) {
            cv::Mat cvFrame = frameRgb->getCvFrame();
            if(cvFrame.channels() == 1) {
                cv::cvtColor(cvFrame, cvFrame, cv::COLOR_GRAY2BGR);
            }

            for(const auto& det : detections->detections) {
                const auto bbox = det.getOuterBoundingBox();
                const int xmin = static_cast<int>(bbox[0] * cvFrame.cols);
                const int ymin = static_cast<int>(bbox[1] * cvFrame.rows);
                const int xmax = static_cast<int>(bbox[2] * cvFrame.cols);
                const int ymax = static_cast<int>(bbox[3] * cvFrame.rows);
                cv::rectangle(cvFrame, cv::Point(xmin, ymin), cv::Point(xmax, ymax), cv::Scalar(0, 255, 0), 2);
            }
            cv::imshow("rgb", cvFrame);
        }

        if(frameFocusedDepth != nullptr) {
            cv::imshow("focused depth", colorizeDepth(frameFocusedDepth->getFrame()));
        }
        if(frameFocusedConf != nullptr) {
            cv::imshow("focused confidence", colorizeConfidence(frameFocusedConf->getFrame()));
        }

        if(cv::waitKey(1) == 'q') {
            pipeline.stop();
            break;
        }
    }

    cv::destroyAllWindows();
    return EXIT_SUCCESS;
}

File: examples/python/Depth/focused_depth.py
#!/usr/bin/env python3
"""Focused depth on RVC4: a fixed ROI or the largest detected object.

Open the printed Visualizer URL. Depth is in the rectified left-camera frame;
color/detections are separate views. --fps requests a camera rate, not a guaranteed
inference rate. The console reports measured depth output FPS after warm-up.
--depth-mode roi leaves the background empty; hold bridges --hold-frames empty
messages using fresh stereo frames; hybrid fills the background with downscaled
EVA stereo and overwrites the selected ROI with neural depth. Hybrid does not hold
missing detections. Neither mode predicts object motion or bridges a stalled detector.
"""

import argparse
import time
from urllib.parse import quote

import depthai as dai
import numpy as np


def parseRoi(text):
    try:
        values = tuple(float(value) for value in text.split(","))
    except ValueError as error:
        raise argparse.ArgumentTypeError("ROI must contain four numbers") from error
    if len(values) != 4 or not (0 <= values[0] < values[2] <= 1 and 0 <= values[1] < values[3] <= 1):
        raise argparse.ArgumentTypeError("ROI must be normalized xmin,ymin,xmax,ymax")
    return values


def main(defaultMode="detector", budget=False):
    parser = argparse.ArgumentParser(description=__doc__)
    parser.add_argument("--mode", choices=("roi", "detector"), default=defaultMode)
    parser.add_argument("--depth-mode", choices=("roi", "hold", "hybrid"), default="roi",
                        help="ROI only; hold ROI through empty detections; or full EVA depth enhanced in the ROI")
    parser.add_argument("--hold-frames", type=int, default=2, help="Consecutive empty detections to bridge in hold mode")
    parser.add_argument("--stereo-size", type=int, nargs=2, default=(384, 240), metavar=("WIDTH", "HEIGHT"),
                        help="Downscaled EVA input size in hybrid mode (default: 384 240)")
    parser.add_argument("--roi", type=parseRoi, default=parseRoi("0.35,0.30,0.65,0.70"))
    parser.add_argument("--model", default="yolov6-nano", help="Object detection model")
    parser.add_argument("--depth-model", choices=("192X120", "288X180", "384X240", "480X300", "576X360"), default="576X360")
    parser.add_argument("--confidence", type=float, default=0.5, help="Detection confidence threshold")
    parser.add_argument("--fps", type=float, default=30.0, help="Requested camera FPS")
    parser.add_argument("--frames", type=int, default=0, help="Stop after N depth frames, including empty frames")
    parser.add_argument("--seconds", type=float, default=0, help="Stop after N seconds (including warm-up)")
    parser.add_argument("--headless", action="store_true", help="Measure without the browser visualizer")
    parser.add_argument("--debug", action="store_true", help="Print per-crop timing")
    parser.add_argument("--webSocketPort", type=int, default=8765)
    parser.add_argument("--httpPort", type=int, default=8082)
    args = parser.parse_args()
    if not np.isfinite(args.fps) or args.fps <= 0 or args.frames < 0 or not np.isfinite(args.seconds) or args.seconds < 0:
        parser.error("FPS must be positive; frames and seconds must be nonnegative")

    if args.hold_frames < 0:
        parser.error("Hold frames must be nonnegative")
    width, height = args.stereo_size
    if width <= 0 or height <= 0 or width % 128 or width > 1280 or height > 800:
        parser.error("Stereo size must be positive, width divisible by 128, and at most 1280x800")

    if not 0 <= args.confidence <= 1:
        parser.error("Confidence must be between zero and one")

    remote = None if args.headless else dai.RemoteConnection(webSocketPort=args.webSocketPort, httpPort=args.httpPort)
    with dai.Pipeline() as pipeline:
        device = pipeline.getDefaultDevice()
        if device.getPlatform() != dai.Platform.RVC4 or not device.isNeuralDepthSupported():
            raise RuntimeError("This example requires an RVC4 with NeuralDepth support")
        depth = pipeline.create(dai.node.Depth)
        depth.setFocusMode(getattr(dai.node.Depth.FocusMode, args.depth_mode.upper()))
        depth.setFocusHoldFrames(args.hold_frames)
        depth.setFocusStereoSize(*args.stereo_size)
        depth.setFocusModels([getattr(dai.DeviceModelZoo, "NEURAL_DEPTH_" + args.depth_model)])
        depth.setFocusSelectionMode(dai.node.Depth.FocusSelectionMode.ALL if budget else dai.node.Depth.FocusSelectionMode.LARGEST)
        depth.setFocusDispatchMode(dai.node.Depth.FocusDispatchMode.TIME_BUDGET if budget else dai.node.Depth.FocusDispatchMode.SINGLE_TIER_PER_FRAME)
        depth.build(args.fps)
        if args.mode == "detector":
            camera = pipeline.create(dai.node.Camera).build(dai.CameraBoardSocket.CAM_A, sensorFps=args.fps)
            detector = pipeline.create(dai.node.DetectionNetwork).build(camera, dai.NNModelDescription(args.model))
            detector.setConfidenceThreshold(args.confidence)
            detector.out.link(depth.inputDetections)
            if remote:
                # Keep the RGB preview small so it does not consume the depth transport budget.
                color = camera.requestOutput((320, 200), type=dai.ImgFrame.Type.RGB888i, fps=args.fps)
                remote.addTopic("color", color, "color")
                remote.addTopic("detections", detector.out, "color")
        else:
            # Timestamp the fixed ROI from the camera, avoiding an unsynchronized host timer.
            left = pipeline.create(dai.node.Camera).build(device.getStereoPairs()[0].left, sensorFps=args.fps)
            trigger = left.requestOutput((64, 40), type=dai.ImgFrame.Type.GRAY8, fps=args.fps)
            script = pipeline.create(dai.node.Script)
            trigger.link(script.inputs["frame"])
            script.setScript(f"""
while True:
    frame = node.io["frame"].get()
    detection = ImgDetection()
    detection.xmin, detection.ymin, detection.xmax, detection.ymax = {args.roi!r}
    detection.confidence = 1.0
    message = ImgDetections()
    message.detections = [detection]
    message.setTimestamp(frame.getTimestamp())
    message.setTimestampDevice(frame.getTimestampDevice())
    message.setSequenceNum(frame.getSequenceNum())
    node.io["detections"].send(message)
""")
            script.outputs["detections"].link(depth.inputDetections)

        # Link detections before accessing the lazy focused output.
        output = depth.focusedDepth
        depthQueue = output.createOutputQueue(maxSize=4, blocking=False)
        debugQueue = depth.focusDebug.createOutputQueue(maxSize=4, blocking=False)
        if remote:
            remote.addTopic("focused depth (left camera)", output, "depth")

        pipeline.start()
        if remote:
            remote.registerPipeline(pipeline)
            wsUrl = quote(f"ws://localhost:{args.webSocketPort}", safe="")
            print(f"Visualizer: http://localhost:{args.httpPort}?ws_url={wsUrl}", flush=True)
        print(f"Mode={args.mode}/{args.depth_mode}, depth model={args.depth_model}, requested camera FPS={args.fps:g}", flush=True)
        if budget:
            print("Time budget is best effort; one crop may exceed the frame period.", flush=True)
        start = time.monotonic()
        first = last = None
        frames = measured = nonempty = 0
        lastReport = start
        try:
            while pipeline.isRunning():
                now = time.monotonic()
                if args.seconds and now - start >= args.seconds:
                    break
                if remote and remote.waitKey(1) == ord("q"):
                    break
                message = depthQueue.tryGet()
                if message is not None:
                    frames += 1
                    if now - start >= 5:
                        first = now if first is None else first
                        last = now
                        measured += 1
                        nonempty += bool(np.any(message.getFrame()))
                for debug in debugQueue.tryGetAll():
                    if args.debug:
                        print(bytes(debug.getData()).decode("utf-8", "replace"), flush=True)
                if now - lastReport >= 2 and measured >= 30:
                    print(f"Depth output: {(measured - 1) / (last - first):.2f} FPS; nonempty={nonempty}/{measured}", flush=True)
                    lastReport = now
                if args.frames and frames >= args.frames:
                    break
                time.sleep(0.001)
        except KeyboardInterrupt:
            pass
        finally:
            if measured > 1:
                print(f"RESULT: {(measured - 1) / (last - first):.2f} depth FPS; {nonempty}/{measured} nonempty frames after warm-up", flush=True)


if __name__ == "__main__":
    main()
